IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -47,71 +47,121 @@ - (instancetype)init
     return self;
 }
 
-- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(MSIDCacheKey *)key
+- (void)storeItem:(id<MSIDJsonSerializable>)item forKey:(MSIDCacheKey *)key
 {
     dispatch_barrier_async(self.queue, ^{
-        [self.cacheObjects setObject:credential forKey:key];
+        NSString *type = [self getItemTypeFromCacheKey:key];
+        
+        if (type)
+        {
+            NSMutableDictionary *items = [self.cacheObjects objectForKey:type];
+            
+            if (!items)
+            {
+                items = [NSMutableDictionary new];
+            }
+            
+            [items setObject:item forKey:key];
+            [self.cacheObjects setObject:items forKey:type];
+        }
     });
 }
 
 - (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
 {
     dispatch_barrier_async(self.queue, ^{
-        for (MSIDCacheKey *key in storageItem.cacheObjects)
+        
+        for (NSString *typeKey in storageItem.cacheObjects)
         {
-            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
-            if (credential)
+            NSMutableDictionary *typeDict = [storageItem.cacheObjects objectForKey:typeKey];
+            NSMutableDictionary *subDict = [self.cacheObjects objectForKey:typeKey];
+            
+            if (!subDict)
             {
-                [self.cacheObjects setObject:credential forKey:key];
+                [self.cacheObjects setObject:typeDict forKey:typeKey];
             }
             else
             {
-                MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, nil, @"Credential is nil for key %@ while merging storage credentials.", MSID_PII_LOG_MASKABLE(key));
+                [subDict addEntriesFromDictionary:typeDict];
+                [self.cacheObjects setObject:subDict forKey:typeKey];
             }
         }
     });
 }
 
-- (void)removeStoredCredentialForKey:(MSIDCacheKey *)key
+- (void)removeStoredItemForKey:(MSIDCacheKey *)key
 {
     dispatch_barrier_sync(self.queue, ^{
-        [self.cacheObjects removeObjectForKey:key];
+        NSString *type = [self getItemTypeFromCacheKey:key];
+        
+        if (type)
+        {
+            NSMutableDictionary *typeDict = [self.cacheObjects msidObjectForKey:type ofClass:[NSDictionary class]];
+            if (typeDict)
+            {
+                [typeDict removeObjectForKey:key];
+                
+                //Update the bucket only if it has one or more items.
+                if ([typeDict count])
+                {
+                    [self.cacheObjects setObject:typeDict forKey:type];
+                }
+                else
+                {
+                    [self.cacheObjects removeObjectForKey:type];
+                }
+            }
+        }
+        
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to get item type from key for stored credential.");
```

Comment:
this will always get logged now? even if type is there?

Reply:
My bad. Addressed.